### ISSUE TITLE
When writing streams of indices, pack content by bits

### ISSUE THREAD

**Issue by Yoric** [binast/binjs-ref#360]
As remarked by Martin Burtscher, our streams of varnums have large amounts of 0 in head positions, and we can improve the packing of sequences of varnums by writing the first bit of all numbers, then the second bit of all numbers, then the third bit, etc.

**Comment by Yoric** [comment]
I wrote a quick hack that modifies streams of varnums as follows:

- pad the number of bytes to make it a multiple of 8;
- group bytes by 8;
- write the first bit of each byte in the group, as our first byte;
- write the second bit...
- ...

before compressing to Brotli.

# Protocol

```
$ cargo run --release --bin binjs_generate_prediction_tables -- --in tests/data/facebook/single/*.js --out /tmp/binjs/facebook-bitpack/dict/
$ cargo run --release --bin binjs_encode -- --in tests/data/facebook/single --out /tmp/binjs/facebook-bitpack/ advanced entropy --dictionary /tmp/binjs/facebook-bitpack/dict/dict.entropy --split-streams
$ cargo run --release --example investigate_streams -- --compare-to-bzip /tmp/binjs/facebook-bitpack/
```

# Results

File | raw | brotli |  size vs. master
-- | -- | -- | --
js | 43134534 | 8016723 |  
binjs | 10854913 | 10773682 |  
  |   |   |  
floats.content | 338496 | 141754 | 0.94103667118086
identifier_names.content | 2525752 | 986819 | 0.90724708032586
list_lengths.content | 1987384 | 651632 | 1.18528327352033
property_keys.content | 1632672 | 1004440 | 0.792914888191928
string_literals.content | 1800600 | 928356 | 0.648808446371176
unsigned_longs.content | 450760 | 121125 | 1.26013046056533

# Bottom line

This seems to give us pretty large gains on identifier names, property keys and string literals, our gang of three. Also gains on floats. Bad on list lengths and unsigned longs, not sure why, but we don't really care all that much about the length of these streams.

# Recommendation

Let's attempt to build a version that roundtrips!

**Comment by Yoric** [comment]
# Experiment 2

Quick hack that modifies all byte streams as follows:

- pad the number of bytes to 8;
- create a bitbuffer;
- walk through all the bytes of the uncompressed stream, write the first bit to the bitbuffer;
- walk through all the bytes of the uncompressed stream, write the second bit to the bitbuffer;
- ...
- walk through all the bytes of the uncompressed stream, write the 7th bit to the bitbuffer;
- compress bitbuffer with brotli as usual.

# Protocol

$ cargo run --release --bin binjs_generate_prediction_tables -- --in tests/data/facebook/single/*.js --out /tmp/binjs/facebook-bitpack-2/dict/
$ cargo run --release --bin binjs_encode -- --in tests/data/facebook/single --out /tmp/binjs/facebook-bitpack-2/ advanced entropy --dictionary /tmp/binjs/facebook-bitpack-2/dict/dict.entropy --split-streams
$ cargo run --release --example investigate_streams -- /tmp/binjs/facebook-bitpack-2/

# Data


File | raw | brotli |  size vs master
-- | -- | -- | --
js | 43134534 | 8016723 | 1
binjs | 2205940 | 2112362 | 0.238527705179692
  |   |   |  
floats.content | 338496 | 22803 | 0.207055298283846
floats.prelude | 127816 | 17797 | 0.245519886324444
identifier_names.content | 2525752 | 51528 | 0.051652844926187
identifier_names.prelude | 87824 | 24578 | 0.464375460539989
identifier_names_len.prelude | 28216 | 19737 | 1.24870302416804
list_lengths.content | 1987384 | 36095 | 0.065568023077328
list_lengths.prelude | 11880 | 13638 | 1.12413452027695
main.entropy | 1611091 | 1601054 | 1
property_keys.content | 1632672 | 41345 | 0.041923332370045
property_keys.prelude | 3151632 | 31516 | 0.031178924645656
property_keys_len.prelude | 222576 | 28026 | 0.193896541465743
string_literals.content | 1800600 | 44843 | 0.051172763851897
string_literals.prelude | 6207104 | 33356 | 0.01688025072405
string_literals_len.prelude | 382104 | 30726 | 0.122594571302034
unsigned_longs.content | 450760 | 30798 | 0.33810888252149
unsigned_longs.prelude | 6400 | 8248 | 1.30156225343222

# Bottom line

This is so good that it is probably a bug.

Let's implement a roundtrip to see if this actually works.
